File: vision-mobile/src/engine/screener.ts
/** Selezione asset: forza relativa vs benchmark + RVOL + trend EMA50.
 * FASE 2: RS percentile non esclude — solo ordina. Regime non filtra.
 */
import { RVOL_HARD_FILTER, RVOL_INTEREST } from "../config";
import { ema, pctReturn, rvol } from "./indicators";
import type { OHLCVBar } from "./types";

export interface ScreenerCandidate {
  symbol: string;
  direction: "long" | "short";
  rs_score: number;
  rvol: number;
  last_price: number;
  rank_score: number;
}

/** Punteggio combinato 0.7*RS + 0.3*RVOL cappato — speculare a rank_score in screener.py. */
export function rankScore(rs: number, rv: number, direction: "long" | "short"): number {
  const strength = direction === "long" ? rs : 1 - rs;
  const volComponent = Math.min(rv / RVOL_INTEREST, 2) / 2;
  return 0.7 * strength + 0.3 * volComponent;
}

function rankPct(raw: Record<string, number>): Record<string, number> {
  const syms = Object.keys(raw);
  const n = syms.length;
  if (n === 0) return {};
  const out: Record<string, number> = {};
  for (const sym of syms) {
    const v = raw[sym];
    let less = 0;
    let equal = 0;
    for (const other of syms) {
      const ov = raw[other];
      if (ov < v) less++;
      else if (ov === v) equal++;
    }
    out[sym] = (less + equal * 0.5) / n;
  }
  return out;
}

export function rsScores(
  data: Record<string, OHLCVBar[]>,
  bench: OHLCVBar[]
): Record<string, number> {
  const benchClose = bench.map((b) => b.close);
  const b20 = pctReturn(benchClose, 20);
  const b60 = pctReturn(benchClose, 60);
  const raw: Record<string, number> = {};
  for (const [sym, bars] of Object.entries(data)) {
    const close = bars.map((b) => b.close);
    const r20 = pctReturn(close, 20) - b20;
    const r60 = pctReturn(close, 60) - b60;
    raw[sym] = 0.5 * r20 + 0.5 * r60;
  }
  if (Object.keys(raw).length === 0) return {};
  return rankPct(raw);
}

/** Direzione da trend EMA50 (RS/regime non escludono). */
export function resolveCandidateDirection(
  _score: number,
  last: number,
  e50: number,
  _longAllowed = true,
  _shortAllowed = true
): "long" | "short" | null {
  if (last > e50) return "long";
  if (last < e50) return "short";
  return null;
}

export function naturalDirection(
  _score: number,
  last: number,
  e50: number
): "long" | "short" | null {
  if (last > e50) return "long";
  if (last < e50) return "short";
  return null;
}

export function classifyCandidates(
  data: Record<string, OHLCVBar[]>,
  scores: Record<string, number>,
  _longAllowed = true,
  _shortAllowed = true,
  rvolHardFilter: boolean | null = null
): ScreenerCandidate[] {
  const hardFilter = rvolHardFilter ?? RVOL_HARD_FILTER;
  const out: ScreenerCandidate[] = [];
  for (const [sym, bars] of Object.entries(data)) {
    const score = scores[sym];
    if (score == null || bars.length < 220) continue;

    const close = bars.map((b) => b.close);
    const last = close[close.length - 1];
    const e50 = ema(close, 50)[close.length - 1];
    const rvSeries = rvol(bars.map((b) => b.volume));
    const rvRaw = rvSeries[rvSeries.length - 1];
    const rv = Number.isFinite(rvRaw) ? rvRaw : 0;

    const direction = resolveCandidateDirection(score, last, e50);
    if (!direction) continue;
    if (hardFilter && rv < RVOL_INTEREST) continue;

    out.push({
      symbol: sym,
      direction,
      rs_score: Math.round(score * 1000) / 1000,
      rvol: Math.round(rv * 100) / 100,
      last_price: last,
      rank_score: Math.round(rankScore(score, rv, direction) * 10000) / 10000,
    });
  }
  out.sort((a, b) => b.rank_score - a.rank_score);
  return out;
}

File: src/backend/engine/screener.py
"""Selezione asset (§3): forza relativa vs benchmark + RVOL + filtro trend."""
import pandas as pd

from config import (
    RS_BOTTOM_PERCENTILE,
    RS_TOP_PERCENTILE,
    RVOL_HARD_FILTER,
    RVOL_INTEREST,
)
from engine.indicators import ema, pct_return, rvol


def rank_score(rs: float, rv: float, direction: str) -> float:
    """Punteggio combinato 0.7*RS + 0.3*RVOL (cappato a 2x la soglia interesse).

    Per gli short la forza RS è speculare (1 - rs): un candidato short è tanto
    più forte quanto più basso è il suo percentile RS; l'RVOL alto premia in
    entrambe le direzioni."""
    strength = rs if direction == "long" else 1.0 - rs
    vol_component = min(rv / RVOL_INTEREST, 2.0) / 2.0
    return 0.7 * strength + 0.3 * vol_component


def rs_scores(data: dict[str, pd.DataFrame], bench: pd.DataFrame) -> dict[str, float]:
    """Score 0..1 = rank percentile della sovraperformance a 20 e 60 giorni vs benchmark."""
    b20 = pct_return(bench["close"], 20)
    b60 = pct_return(bench["close"], 60)
    raw: dict[str, float] = {}
    for sym, df in data.items():
        r20 = pct_return(df["close"], 20) - b20
        r60 = pct_return(df["close"], 60) - b60
        raw[sym] = 0.5 * r20 + 0.5 * r60
    if not raw:
        return {}
    s = pd.Series(raw)
    ranks = s.rank(pct=True)
    return {sym: float(ranks[sym]) for sym in raw}


def resolve_candidate_direction(
    score: float,
    last: float,
    e50: float,
    long_allowed: bool = True,
    short_allowed: bool = True,
) -> str | None:
    """Direzione da trend EMA50 (FASE 2: RS non esclude più).

    long_allowed/short_allowed restano in firma per retrocompatibilità ma
    non filtrano: il regime è solo banner informativo.
    """
    del score, long_allowed, short_allowed  # RS/regime non escludono
    if last > e50:
        return "long"
    if last < e50:
        return "short"
    return None


def natural_direction(score: float, last: float, e50: float) -> str | None:
    """Direzione da trend EMA50; RS usato solo per ordinamento a valle."""
    del score
    if last > e50:
        return "long"
    if last < e50:
        return "short"
    return None


def classify_candidates(
    data: dict[str, pd.DataFrame],
    scores: dict[str, float],
    long_allowed: bool = True,
    short_allowed: bool = True,
    rvol_hard_filter: bool | None = None,
) -> list[dict]:
    """Candidati per trend EMA50, ORDINATI per rank RS (non esclusi da percentile).

    FASE 2: il gate RS percentile non esclude più nulla — il rank_score ordina
    l'attenzione. long_allowed/short_allowed ignorati (regime = banner).
    Con rvol_hard_filter=True i candidati con RVOL < RVOL_INTEREST cadono.
    """
    del long_allowed, short_allowed
    if rvol_hard_filter is None:
        rvol_hard_filter = RVOL_HARD_FILTER
    out: list[dict] = []
    for sym, df in data.items():
        score = scores.get(sym)
        if score is None or len(df) < 220:
            continue
        close = df["close"]
        last = float(close.iloc[-1])
        e50 = float(ema(close, 50).iloc[-1])
        rv_series = rvol(df["volume"])
        rv = float(rv_series.iloc[-1]) if pd.notna(rv_series.iloc[-1]) else 0.0

        direction = resolve_candidate_direction(score, last, e50)
        if direction is None:
            continue
        if rvol_hard_filter and rv < RVOL_INTEREST:
            continue
        out.append(
            {
                "symbol": sym,
                "direction": direction,
                "rs_score": round(score, 3),
                "rvol": round(rv, 2),
                "last_price": last,
                "rank_score": round(rank_score(score, rv, direction), 4),
            }
        )
    out.sort(key=lambda c: -c["rank_score"])
    return out
